Batch Exporter: AssertionError when starting export twice

This is due to how the background process is implemented. The plugin does not test for running processes nor cancels them if necessary. The correct way to fix this issue is by implementing #914 

```
Traceback (most recent call last):
  File "/home/papr/work/pupil/pupil_src/launchables/player.py", line 441, in player
    p.recent_events(events)
  File "/home/papr/work/pupil/pupil_src/shared_modules/batch_exporter.py", line 190, in recent_events
    self.workers[i].start()
  File "/usr/lib/python3.5/multiprocessing/process.py", line 99, in start
    assert self._popen is None, 'cannot start a process twice'
AssertionError: cannot start a process twice
```

--- a/pupil_src/launchables/player.py
+++ b/pupil_src/launchables/player.py
@@ -94,7 +94,7 @@ def player(rec_dir, ipc_pub_url, ipc_sub_url,
         from offline_surface_tracker import Offline_Surface_Tracker
         # from marker_auto_trim_marks import Marker_Auto_Trim_Marks
         from fixation_detector import Offline_Fixation_Detector
-        # from batch_exporter import Batch_Exporter
+        from batch_exporter import Batch_Exporter
         from log_display import Log_Display
         from annotations import Annotation_Player
         from raw_data_exporter import Raw_Data_Exporter
@@ -109,7 +109,7 @@ def player(rec_dir, ipc_pub_url, ipc_sub_url,
         system_plugins = [Log_Display, Seek_Control, Plugin_Manager, System_Graphs]
         user_plugins = [Vis_Circle, Vis_Fixation, Vis_Polyline, Vis_Light_Points,
                         Vis_Cross, Vis_Watermark, Vis_Eye_Video_Overlay, Vis_Scan_Path,
-                        Offline_Fixation_Detector,
+                        Offline_Fixation_Detector, Batch_Exporter,
                         Video_Export_Launcher, Offline_Surface_Tracker, Raw_Data_Exporter,
                         Annotation_Player, Log_History,
                         Pupil_From_Recording, Offline_Pupil_Detection, Gaze_From_Recording,

--- a/pupil_src/launchables/player.py
+++ b/pupil_src/launchables/player.py
@@ -103,7 +103,7 @@ def player(rec_dir, ipc_pub_url, ipc_sub_url,
         from gaze_producers import Gaze_From_Recording, Offline_Calibration
         from system_graphs import System_Graphs
 
-        assert VersionFormat(pyglui_version) >= VersionFormat('1.10'), 'pyglui out of date, please upgrade to newest version'
+        assert VersionFormat(pyglui_version) >= VersionFormat('1.11'), 'pyglui out of date, please upgrade to newest version'
 
         runtime_plugins = import_runtime_plugins(os.path.join(user_dir, 'plugins'))
         system_plugins = [Log_Display, Seek_Control, Plugin_Manager, System_Graphs]

--- a/pupil_src/launchables/player.py
+++ b/pupil_src/launchables/player.py
@@ -94,7 +94,7 @@ def player(rec_dir, ipc_pub_url, ipc_sub_url,
         from offline_surface_tracker import Offline_Surface_Tracker
         # from marker_auto_trim_marks import Marker_Auto_Trim_Marks
         from fixation_detector import Offline_Fixation_Detector
-        from batch_exporter import Batch_Exporter
+        from batch_exporter import Batch_Exporter, Batch_Export
         from log_display import Log_Display
         from annotations import Annotation_Player
         from raw_data_exporter import Raw_Data_Exporter
@@ -106,7 +106,7 @@ def player(rec_dir, ipc_pub_url, ipc_sub_url,
         assert VersionFormat(pyglui_version) >= VersionFormat('1.11'), 'pyglui out of date, please upgrade to newest version'
 
         runtime_plugins = import_runtime_plugins(os.path.join(user_dir, 'plugins'))
-        system_plugins = [Log_Display, Seek_Control, Plugin_Manager, System_Graphs]
+        system_plugins = [Log_Display, Seek_Control, Plugin_Manager, System_Graphs, Batch_Export]
         user_plugins = [Vis_Circle, Vis_Fixation, Vis_Polyline, Vis_Light_Points,
                         Vis_Cross, Vis_Watermark, Vis_Eye_Video_Overlay, Vis_Scan_Path,
                         Offline_Fixation_Detector, Batch_Exporter,

--- a/pupil_src/shared_modules/background_helper.py
+++ b/pupil_src/shared_modules/background_helper.py
@@ -33,6 +33,7 @@ class Task_Proxy(object):
         wrapper_args = [pipe_send, self._should_terminate_flag, generator]
         wrapper_args.extend(args)
         self.process = mp.Process(target=self._wrapper, name=name, args=wrapper_args, kwargs=kwargs)
+        self.process.daemon = True
         self.process.start()
         self.pipe = pipe_recv
 
@@ -61,13 +62,17 @@ class Task_Proxy(object):
                 datum = self.pipe.recv()
             except EOFError:
                 logger.debug("Process canceled be user.")
+                self._canceled = True
+                self.process = None
                 return
             else:
                 if isinstance(datum, StopIteration):
                     self._completed = True
+                    self.process = None
                     return
                 elif isinstance(datum, EarlyCancellationError):
                     self._canceled = True
+                    self.process = None
                     return
                 elif isinstance(datum, Exception):
                     raise datum

--- a/pupil_src/shared_modules/batch_exporter.py
+++ b/pupil_src/shared_modules/batch_exporter.py
@@ -55,10 +55,9 @@ class Batch_Exporter(Analysis_Plugin_Base):
     icon_chr = chr(0xec05)
     icon_font = 'pupil_icons'
 
-    def __init__(self, g_pool, source_dir='~/', destination_dir='~/'):
+    def __init__(self, g_pool, source_dir='~/work/pupil/recordings/BATCH', destination_dir='~/'):
         super().__init__(g_pool)
 
-        self._requires_ui_update = False
         self.available_exports = []
         self.queued_exports = []
         self.active_exports = []
@@ -91,15 +90,7 @@ class Batch_Exporter(Analysis_Plugin_Base):
         self.menu.append(ui.Button('Export selected', self.queue_selected))
         self.menu.append(ui.Button('Clear search results', self._clear_avail))
 
-        self._mark_outdated_ui()
-
-    def _mark_outdated_ui(self):
-        self._requires_ui_update = True
-
-    def gl_display(self):
-        if self._requires_ui_update:
-            self._update_ui()
-            self._requires_ui_update = False
+        self._update_ui()
 
     def _update_ui(self):
         del self.menu.elements[7:]
@@ -114,7 +105,7 @@ class Batch_Exporter(Analysis_Plugin_Base):
                             self.queued_exports.remove(qd)
                             self.available_exports.append({'source': qd['source'], 'selected': True})
                             self._update_avail_recs_menu()
-                            self._mark_outdated_ui()
+                            self._update_ui()
                     return cancel
                 self.menu.append(ui.Button('Cancel', cancel_qd(queued), outer_label=queued['dest']))
 
@@ -183,7 +174,7 @@ class Batch_Exporter(Analysis_Plugin_Base):
 
     def _clear_previous(self):
         del self.previous_exports[:]
-        self._mark_outdated_ui()
+        self._update_ui()
 
     def _clear_avail(self):
         del self.available_exports[:]
@@ -224,7 +215,7 @@ class Batch_Exporter(Analysis_Plugin_Base):
                 self.available_exports.remove(avail)
                 self.queued_exports.append(export)
         self._update_avail_recs_menu()
-        self._mark_outdated_ui()
+        self._update_ui()
 
     def start_export(self, queued):
         user_dir = self.g_pool.user_dir
@@ -239,6 +230,7 @@ class Batch_Exporter(Analysis_Plugin_Base):
         process.progress = 0
         self.active_exports.append(process)
         self.queued_exports.remove(queued)
+        self._update_ui()
 
     def recent_events(self, events):
         if self.search_task:
@@ -264,7 +256,7 @@ class Batch_Exporter(Analysis_Plugin_Base):
             if process.canceled or process.completed:
                 self.active_exports.remove(process)
                 self.previous_exports.append(process)
-                self._mark_outdated_ui()
+                self._update_ui()
 
         # Add queued exports to active queue
         for queued in self.queued_exports[:self.worker_count - len(self.active_exports)]:

--- a/pupil_src/shared_modules/batch_exporter.py
+++ b/pupil_src/shared_modules/batch_exporter.py
@@ -29,7 +29,7 @@ if __name__ == '__main__':
 
 from plugin import Analysis_Plugin_Base
 
-from exporter import export
+from exporter import export as export_function
 from player_methods import is_pupil_rec_dir
 
 
@@ -55,9 +55,10 @@ class Batch_Exporter(Analysis_Plugin_Base):
     icon_chr = chr(0xec05)
     icon_font = 'pupil_icons'
 
-    def __init__(self, g_pool, source_dir='~/work/pupil/recordings/BATCH', destination_dir='~/'):
+    def __init__(self, g_pool, source_dir='~/', destination_dir='~/'):
         super().__init__(g_pool)
 
+        self._requires_ui_update = False
         self.available_exports = []
         self.queued_exports = []
         self.active_exports = []
@@ -68,6 +69,9 @@ class Batch_Exporter(Analysis_Plugin_Base):
         self.search_task = None
         self.worker_count = cpu_count() - 1
         logger.info("Using a maximum of {} CPUs to process visualizations in parallel...".format(cpu_count() - 1))
+
+    def get_init_dict(self):
+        return {'source_dir': self.source_dir, 'destination_dir': self.destination_dir}
 
     def init_ui(self):
         self.add_menu()
@@ -84,10 +88,18 @@ class Batch_Exporter(Analysis_Plugin_Base):
         self.menu.append(self.avail_recs_menu)
 
         self.menu.append(ui.Text_Input('destination_dir', self, label='Destination Directory', setter=self.set_dest_dir))
-        self.menu.append(ui.Button('Export selected', self.export_selected))
+        self.menu.append(ui.Button('Export selected', self.queue_selected))
         self.menu.append(ui.Button('Clear search results', self._clear_avail))
 
-        self._update_ui()
+        self._mark_outdated_ui()
+
+    def _mark_outdated_ui(self):
+        self._requires_ui_update = True
+
+    def gl_display(self):
+        if self._requires_ui_update:
+            self._update_ui()
+            self._requires_ui_update = False
 
     def _update_ui(self):
         del self.menu.elements[7:]
@@ -96,9 +108,15 @@ class Batch_Exporter(Analysis_Plugin_Base):
             self.menu.append(ui.Separator())
             self.menu.append(ui.Info_Text('Queued exports:'))
             for queued in self.queued_exports[::-1]:
-                def prio_qd():
-                    pass
-                self.menu.append(ui.Button('Prioritize', prio_qd, outer_label=queued['dest']))
+                def cancel_qd(qd):
+                    def cancel():
+                        if qd in self.queued_exports:
+                            self.queued_exports.remove(qd)
+                            self.available_exports.append({'source': qd['source'], 'selected': True})
+                            self._update_avail_recs_menu()
+                            self._mark_outdated_ui()
+                    return cancel
+                self.menu.append(ui.Button('Cancel', cancel_qd(queued), outer_label=queued['dest']))
 
         if self.active_exports:
             self.menu.append(ui.Separator())
@@ -165,7 +183,7 @@ class Batch_Exporter(Analysis_Plugin_Base):
 
     def _clear_previous(self):
         del self.previous_exports[:]
-        self._update_ui()
+        self._mark_outdated_ui()
 
     def _clear_avail(self):
         del self.available_exports[:]
@@ -179,7 +197,7 @@ class Batch_Exporter(Analysis_Plugin_Base):
         else:
             self.avail_recs_menu.append(ui.Info_Text('No recordings available yet. Use Search to find recordings.'))
 
-    def export_selected(self):
+    def queue_selected(self):
         for avail in self.available_exports[:]:
             if avail['selected']:
                 try:
@@ -206,7 +224,21 @@ class Batch_Exporter(Analysis_Plugin_Base):
                 self.available_exports.remove(avail)
                 self.queued_exports.append(export)
         self._update_avail_recs_menu()
-        self._update_ui()
+        self._mark_outdated_ui()
+
+    def start_export(self, queued):
+        user_dir = self.g_pool.user_dir
+        plugins = self.g_pool.plugins.get_initializers()
+        out_file_path = queued['dest']
+        args = (queued['source'], user_dir, self.g_pool.min_data_confidence, None,
+                None, plugins, out_file_path, {})
+        process = bh.Task_Proxy('Pupil Export {}'.format(out_file_path), export_function, args=args)
+        process.out_file_path = out_file_path
+        process.frames_to_export = queued['frames_to_export']
+        process.status = ''
+        process.progress = 0
+        self.active_exports.append(process)
+        self.queued_exports.remove(queued)
 
     def recent_events(self, events):
         if self.search_task:
@@ -220,6 +252,29 @@ class Batch_Exporter(Analysis_Plugin_Base):
                 self.search_button.outer_label = ''
                 self.search_button.label = 'Search'
 
+        for process in self.active_exports[:]:
+            recent = [d for d in process.fetch()]
+            if recent:
+                process.status, process.progress = recent[-1]
+            # Update status if process has been canceled
+            if process.canceled:
+                process.status = 'Export has been canceled.'
+            # Move process from `active` to `previous` exports list after
+            # completion or cancelation
+            if process.canceled or process.completed:
+                self.active_exports.remove(process)
+                self.previous_exports.append(process)
+                self._mark_outdated_ui()
+
+        # Add queued exports to active queue
+        for queued in self.queued_exports[:self.worker_count - len(self.active_exports)]:
+            self.start_export(queued)
+
+    def cleanup(self):
+        if self.search_task:
+            self.search_task.cancel()
+        for process in self.active_exports:
+            process.cancel()
 
 def main():
 

--- a/pupil_src/shared_modules/batch_exporter.py
+++ b/pupil_src/shared_modules/batch_exporter.py
@@ -27,7 +27,7 @@ if __name__ == '__main__':
     pupil_base_dir = os.path.abspath(__file__).rsplit('pupil_src', 1)[0]
     sys.path.append(os.path.join(pupil_base_dir, 'pupil_src', 'shared_modules'))
 
-from plugin import Analysis_Plugin_Base
+from plugin import Analysis_Plugin_Base, System_Plugin_Base
 
 from exporter import export as export_function
 from player_methods import is_pupil_rec_dir
@@ -48,10 +48,115 @@ def get_recording_dirs(data_dir):
                 yield joined
 
 
+class Batch_Export(System_Plugin_Base):
+    """Sub plugin that manages a single batch export"""
+    uniqueness = 'not_unique'
+    menu_icon_order = {'in_queue': 0.8, 'active': 0.7, 'finished': 0.9}
+    icon_font = 'pupil_icons'
+    icon_chr = chr(0xe2c4)  # character shown in menu icon
+
+    def __init__(self, g_pool, rec_dir, out_file_path, frames_to_export):
+        super().__init__(g_pool)
+        self.rec_dir = rec_dir
+        self.out_file_path = out_file_path
+        self.plugins = self.g_pool.plugins.get_initializers()
+        self.process = None
+        self.status = 'In queue'
+        self.progress = 0
+        self.frames_to_export = frames_to_export
+        self.in_queue = True
+        self._accelerate = True
+        self.notify_all({'subject': 'batch_export.queued', 'out_file_path': self.out_file_path})
+
+    def init_ui(self):
+        self.add_menu()
+        # uniqueness = 'not_unique' -> Automatic `Close` button
+        # -> Rename to `Cancel`
+        export_name = os.path.split(self.out_file_path)[-1]
+        self.menu.label = 'Batch Export {}'.format(export_name)
+        self.menu[0].label = 'Cancel'
+        self.menu_icon.indicator_start = 0.
+        self.menu_icon.indicator_stop = 0.1
+        self.menu_icon.order = self.menu_icon_order['in_queue']
+        self.menu_icon.tooltip = export_name
+        self.menu.append(ui.Text_Input('rec_dir', self, label='Recording', setter=lambda x: None))
+        self.menu.append(ui.Text_Input('out_file_path', self, label='Output', setter=lambda x: None))
+        self.menu.append(ui.Text_Input('status', self, label='Status', setter=lambda x: None))
+        progress_bar = ui.Slider('progress', self, min=0, max=self.frames_to_export, label='Progress')
+        progress_bar.read_only = True
+        self.menu.append(progress_bar)
+
+    def recent_events(self, events):
+        if self.process:
+            recent = [d for d in self.process.fetch()]
+            if recent:
+                self.status, self.progress = recent[-1]
+
+            # Update status if process has been canceled or completed
+            if self.process.canceled:
+                self.process = None
+                self.status = 'Export has been canceled.'
+                self.notify_all({'subject': 'batch_export.canceled', 'out_file_path': self.out_file_path})
+                self.menu_icon.order = self.menu_icon_order['finished']
+                self.menu[0].label = 'Close'  # change button label back to close
+            elif self.process.completed:
+                self.process = None
+                self.notify_all({'subject': 'batch_export.completed', 'out_file_path': self.out_file_path})
+                self.menu_icon.order = self.menu_icon_order['finished']
+                self.menu[0].label = 'Close'  # change button label back to close
+
+        if self.in_queue:
+            if self._accelerate:
+                self.menu_icon.indicator_start += 0.01
+                self.menu_icon.indicator_stop += 0.02
+            else:
+                self.menu_icon.indicator_start += 0.02
+                self.menu_icon.indicator_stop += 0.01
+            d = abs(self.menu_icon.indicator_start - self.menu_icon.indicator_stop)
+            if self._accelerate and d > .5:
+                self._accelerate = False
+            elif not self._accelerate and d < .1:
+                self._accelerate = True
+        else:
+            self.menu_icon.indicator_start = 0.
+            self.menu_icon.indicator_stop = max(1, self.progress) / self.frames_to_export
+
+    def on_notify(self, n):
+        if n['subject'] == 'batch_export.should_start' and n['out_file_path'] == self.out_file_path:
+            self.init_export()
+        if n['subject'] == 'batch_export.should_cancel' and n.get('out_file_path', self.out_file_path) == self.out_file_path:
+            self.cancel_export()
+            self.menu_icon.order = self.menu_icon_order['finished']
+            if n.get('remove_menu', False):
+                self.alive = False
+
+    def init_export(self):
+        self.in_queue = False
+        self.menu_icon.order = self.menu_icon_order['active']
+        args = (self.rec_dir, self.g_pool.user_dir, self.g_pool.min_data_confidence,
+                None, None, self.plugins, self.out_file_path, {})
+        self.process = bh.Task_Proxy('Pupil Batch Export {}'.format(self.out_file_path), export_function, args=args)
+        self.notify_all({'subject': 'batch_export.started', 'out_file_path': self.out_file_path})
+
+    def cancel_export(self):
+        if self.process:
+            self.process.cancel()
+        self.notify_all({'subject': 'batch_export.canceled', 'out_file_path': self.out_file_path})
+
+    def get_init_dict(self):
+        # do not be session persistent
+        raise NotImplementedError()
+
+    def deinit_ui(self):
+        self.remove_menu()
+
+    def cleanup(self):
+        self.cancel_export()
+        self.notify_all({'subject': 'batch_export.removed', 'out_file_path': self.out_file_path})
+
+
 class Batch_Exporter(Analysis_Plugin_Base):
-    """docstring for Batch_Exporter
-    this plugin can export videos in a seperate process using exporter
-    """
+    """The Batch_Exporter searches for available recordings and exports them to a common location"""
     icon_chr = chr(0xec05)
     icon_font = 'pupil_icons'
 
@@ -61,7 +166,6 @@ class Batch_Exporter(Analysis_Plugin_Base):
         self.available_exports = []
         self.queued_exports = []
         self.active_exports = []
-        self.previous_exports = []
         self.destination_dir = os.path.expanduser(destination_dir)
         self.source_dir = os.path.expanduser(source_dir)
 
@@ -89,56 +193,9 @@ class Batch_Exporter(Analysis_Plugin_Base):
         self.menu.append(ui.Text_Input('destination_dir', self, label='Destination Directory', setter=self.set_dest_dir))
         self.menu.append(ui.Button('Export selected', self.queue_selected))
         self.menu.append(ui.Button('Clear search results', self._clear_avail))
+        self.menu.append(ui.Separator())
+        self.menu.append(ui.Button('Cancel all exports', self.cancel_all))
 
-        self._update_ui()
-
-    def _update_ui(self):
-        del self.menu.elements[7:]
-
-        if self.queued_exports:
-            self.menu.append(ui.Separator())
-            self.menu.append(ui.Info_Text('Queued exports:'))
-            for queued in self.queued_exports[::-1]:
-                def cancel_qd(qd):
-                    def cancel():
-                        if qd in self.queued_exports:
-                            self.queued_exports.remove(qd)
-                            self.available_exports.append({'source': qd['source'], 'selected': True})
-                            self._update_avail_recs_menu()
-                            self._update_ui()
-                    return cancel
-                self.menu.append(ui.Button('Cancel', cancel_qd(queued), outer_label=queued['dest']))
-
-        if self.active_exports:
-            self.menu.append(ui.Separator())
-
-            for job in self.active_exports[::-1]:
-                submenu = ui.Growing_Menu('Active: {}'.format(job.out_file_path))
-                submenu.append(ui.Text_Input('status', job, label='Status', setter=lambda x: None))
-                progress_bar = ui.Slider('progress', job, min=0, max=job.frames_to_export, label='Progress')
-                progress_bar.read_only = True
-                submenu.append(progress_bar)
-                submenu.append(ui.Button('Cancel', job.cancel))
-                self.menu.append(submenu)
-
-        if self.previous_exports:
-            self.menu.append(ui.Separator())
-
-            for job in self.previous_exports[::-1]:
-                if job.completed:
-                    status = 'Completed'
-                elif job.canceled:
-                    status = 'Canceled'
-                else:
-                    status = 'Previous'
-                submenu = ui.Growing_Menu('{}: {}'.format(status, job.out_file_path))
-                progress_bar = ui.Slider('progress', job, min=0, max=job.frames_to_export, label='Progress')
-                progress_bar.read_only = True
-                submenu.append(progress_bar)
-                submenu.collapsed = True
-                self.menu.append(submenu)
-
-            self.menu.append(ui.Button('Clear previous exports', self._clear_previous))
 
     def deinit_ui(self):
         self.menu.remove(self.avail_recs_menu)
@@ -172,10 +229,6 @@ class Batch_Exporter(Analysis_Plugin_Base):
             logger.warning('"{}" is not a directory'.format(new_dir))
             return
 
-    def _clear_previous(self):
-        del self.previous_exports[:]
-        self._update_ui()
-
     def _clear_avail(self):
         del self.available_exports[:]
         self._update_avail_recs_menu()
@@ -203,34 +256,34 @@ class Batch_Exporter(Analysis_Plugin_Base):
                 out_name = rec_session+"_"+rec_dir+".mp4"
                 out_file_path = os.path.join(self.destination_dir, out_name)
 
-                if (out_file_path in (e['dest'] for e in self.queued_exports) or
-                        out_file_path in (e.out_file_path for e in self.active_exports)):
+                if out_file_path in self.queued_exports or out_file_path in self.active_exports:
                     logger.error("This export setting would try to save {} at least twice please rename dirs to prevent this. Skipping recording.".format(out_file_path))
                     continue
-                if out_file_path in (e.out_file_path for e in self.previous_exports):
-                    logger.error("This export setting would the previous export {}. Please clear the previous exports if you want to overwrite it.".format(out_file_path))
-                    continue
 
-                export = {'source': avail['source'], 'dest': out_file_path, 'frames_to_export': frames_to_export}
+                self.notify_all({'subject': 'start_plugin', 'name': 'Batch_Export',
+                                'args': {'out_file_path': out_file_path,
+                                         'rec_dir': avail['source'],
+                                         'frames_to_export': frames_to_export}})
                 self.available_exports.remove(avail)
-                self.queued_exports.append(export)
+                self.queued_exports.append(out_file_path)
         self._update_avail_recs_menu()
-        self._update_ui()
 
     def start_export(self, queued):
-        user_dir = self.g_pool.user_dir
-        plugins = self.g_pool.plugins.get_initializers()
-        out_file_path = queued['dest']
-        args = (queued['source'], user_dir, self.g_pool.min_data_confidence, None,
-                None, plugins, out_file_path, {})
-        process = bh.Task_Proxy('Pupil Export {}'.format(out_file_path), export_function, args=args)
-        process.out_file_path = out_file_path
-        process.frames_to_export = queued['frames_to_export']
-        process.status = ''
-        process.progress = 0
-        self.active_exports.append(process)
+        self.notify_all({'subject': 'batch_export.should_start', 'out_file_path': queued})
+        self.active_exports.append(queued)
         self.queued_exports.remove(queued)
-        self._update_ui()
+
+    def on_notify(self, n):
+        if n['subject'] in ('batch_export.canceled', 'batch_export.completed'):
+            if n['out_file_path'] in self.queued_exports:
+                self.queued_exports.remove(n['out_file_path'])
+            if n['out_file_path'] in self.active_exports:
+                self.active_exports.remove(n['out_file_path'])
+
+        # Add queued exports to active queue
+        for queued in self.queued_exports[:self.worker_count - len(self.active_exports)]:
+            self.start_export(queued)
+
 
     def recent_events(self, events):
         if self.search_task:
@@ -244,29 +297,13 @@ class Batch_Exporter(Analysis_Plugin_Base):
                 self.search_button.outer_label = ''
                 self.search_button.label = 'Search'
 
-        for process in self.active_exports[:]:
-            recent = [d for d in process.fetch()]
-            if recent:
-                process.status, process.progress = recent[-1]
-            # Update status if process has been canceled
-            if process.canceled:
-                process.status = 'Export has been canceled.'
-            # Move process from `active` to `previous` exports list after
-            # completion or cancelation
-            if process.canceled or process.completed:
-                self.active_exports.remove(process)
-                self.previous_exports.append(process)
-                self._update_ui()
-
-        # Add queued exports to active queue
-        for queued in self.queued_exports[:self.worker_count - len(self.active_exports)]:
-            self.start_export(queued)
+    def cancel_all(self):
+        self.notify_all({'subject': 'batch_export.should_cancel', 'remove_menu': True})
 
     def cleanup(self):
+        self.cancel_all()
         if self.search_task:
             self.search_task.cancel()
-        for process in self.active_exports:
-            process.cancel()
 
 def main():
 

--- a/pupil_src/shared_modules/batch_exporter.py
+++ b/pupil_src/shared_modules/batch_exporter.py
@@ -9,13 +9,14 @@ See COPYING and COPYING.LESSER for license details.
 ---------------------------------------------------------------------------~(*)
 '''
 
-import numpy as np
 from pyglui import ui
 import os
 import sys
 import time
 
 from ctypes import c_bool, c_int
+from multiprocessing import cpu_count
+import background_helper as bh
 
 import logging
 logger = logging.getLogger(__name__)
@@ -26,7 +27,6 @@ if __name__ == '__main__':
     sys.path.append(os.path.join(pupil_base_dir, 'pupil_src', 'shared_modules'))
 
 from plugin import Analysis_Plugin_Base
-from video_export_launcher import mp, Export_Process
 
 from exporter import export
 from player_methods import is_pupil_rec_dir
@@ -38,14 +38,13 @@ def get_recording_dirs(data_dir):
         - all folders within will be checked for necessary files
         - in order to make a visualization
     '''
-    filtered_recording_dirs = []
     if is_pupil_rec_dir(data_dir):
-        filtered_recording_dirs.append(data_dir)
+        yield data_dir
     for root, dirs, files in os.walk(data_dir):
-        filtered_recording_dirs += [os.path.join(root, d) for d in dirs
-                                    if not d.startswith(".") and is_pupil_rec_dir(os.path.join(root, d))]
-    logger.debug("Filtered Recording Dirs: {}".format(filtered_recording_dirs))
-    return filtered_recording_dirs
+        for d in dirs:
+            joined = os.path.join(root, d)
+            if not d.startswith(".") and is_pupil_rec_dir(joined):
+                yield joined
 
 
 class Batch_Exporter(Analysis_Plugin_Base):
@@ -55,68 +54,105 @@ class Batch_Exporter(Analysis_Plugin_Base):
     icon_chr = chr(0xec05)
     icon_font = 'pupil_icons'
 
-    def __init__(self, g_pool):
+    def __init__(self, g_pool, source_dir='~/work/pupil/recordings/BATCH', destination_dir='~/'):
         super().__init__(g_pool)
 
-        self.exports = []
-        self.new_exports = []
+        self.available_exports = []
+        self.queued_exports = []
         self.active_exports = []
-        default_path = os.path.expanduser('~/work/pupil/recordings/BATCH')
-        self.destination_dir = default_path
-        self.source_dir = default_path
+        self.previous_exports = []
+        self.destination_dir = os.path.expanduser(destination_dir)
+        self.source_dir = os.path.expanduser(source_dir)
 
-        self.run = False
-        self.workers = [None for x in range(mp.cpu_count() - 1)]
-        logger.info("Using a maximum of {} CPUs to process visualizations in parallel...".format(mp.cpu_count() - 1))
+        self.search_task = None
+        self.worker_count = cpu_count() - 1
+        logger.info("Using a maximum of {} CPUs to process visualizations in parallel...".format(cpu_count() - 1))
 
     def init_ui(self):
         self.add_menu()
-        # initialize the menu
         self.menu.label = 'Batch Export Recordings'
-        # load the configuration of last session
-        # add menu to the window
+
+        self.menu.append(ui.Info_Text('Search will walk through the source direcotry recursively and detect available Pupil recordings.'))
+        self.menu.append(ui.Text_Input('source_dir', self, label='Source Directory', setter=self.set_src_dir))
+
+        self.search_button = ui.Button('Search', self.detect_recordings)
+        self.menu.append(self.search_button)
+
+        self.avail_recs_menu = ui.Growing_Menu('Available Recordings')
+        self._update_avail_recs_menu()
+        self.menu.append(self.avail_recs_menu)
+
+        self.menu.append(ui.Text_Input('destination_dir', self, label='Destination Directory', setter=self.set_dest_dir))
+        self.menu.append(ui.Button('Export selected', self.export_selected))
+        self.menu.append(ui.Button('Clear search results', self._clear_avail))
+        self.menu.append(ui.Separator())
+
         self._update_ui()
 
     def _update_ui(self):
-        self.menu.elements[:] = []
-        self.menu.append(ui.Text_Input('source_dir', self, label='Recording Source Directory', setter=self.set_src_dir))
-        self.menu.append(ui.Text_Input('destination_dir', self, label='Recording Destination Directory', setter=self.set_dest_dir))
-        self.menu.append(ui.Button('Start Export', self.start))
+        del self.menu.elements[7:]
 
-        for idx, job in enumerate(self.exports[::-1]):
-            submenu = ui.Growing_Menu("Export Job {}: '{}'".format(idx, job.out_file_path))
-            progress_bar = ui.Slider('Progress', getter=job.status, min=0, max=job.frames_to_export.value)
-            progress_bar.read_only = True
-            submenu.append(progress_bar)
-            submenu.append(ui.Button('Cancel', job.cancel))
-            self.menu.append(submenu)
-        if not self.exports:
-            self.menu.append(ui.Info_Text('Please select a Recording Source directory from which to pull all recordings for the batch export.'))
+        for queued in self.queued_exports[::-1]:
+            def prio_qd():
+                pass
+            self.menu.append(ui.Button('Prioritize', prio_qd, outer_label=queued['dest']))
+
+        if self.active_exports:
+            self.menu.append(ui.Separator())
+
+            for job in self.active_exports[::-1]:
+                submenu = ui.Growing_Menu('Active: {}'.format(job.out_file_path))
+                submenu.append(ui.Text_Input('status', job, label='Status', setter=lambda x: None))
+                progress_bar = ui.Slider('progress', job, min=0, max=job.frames_to_export, label='Progress')
+                progress_bar.read_only = True
+                submenu.append(progress_bar)
+                submenu.append(ui.Button('Cancel', job.cancel))
+                self.menu.append(submenu)
+
+        if self.previous_exports:
+            self.menu.append(ui.Separator())
+
+            for job in self.previous_exports[::-1]:
+                if job.completed:
+                    status = 'Completed'
+                elif job.canceled:
+                    status = 'Canceled'
+                else:
+                    status = 'Previous'
+                submenu = ui.Growing_Menu('{}: {}'.format(status, job.out_file_path))
+                progress_bar = ui.Slider('progress', job, min=0, max=job.frames_to_export, label='Progress')
+                progress_bar.read_only = True
+                submenu.append(progress_bar)
+                submenu.collapsed = True
+                self.menu.append(submenu)
+
+            self.menu.append(ui.Button('Clear previous exports', self._clear_previous))
 
     def deinit_ui(self):
+        self.menu.remove(self.avail_recs_menu)
+        self.avail_recs_menu = None
         self.remove_menu()
 
+    def detect_recordings(self):
+        if self.search_task:
+            self.search_task.cancel()
+            self.search_task = None
+            self.search_button.outer_label = ''
+            self.search_button.label = 'Search'
+        else:
+            self.search_button.outer_label = 'Searching...'
+            self.search_button.label = 'Cancel'
+            self.search_task = bh.Task_Proxy('Searching recordings in {}'.format(self.source_dir), get_recording_dirs, args=[self.source_dir])
+
     def set_src_dir(self, new_dir):
-        new_dir = new_dir
-        self.new_exports = []
-        self.exports = []
         new_dir = os.path.expanduser(new_dir)
         if os.path.isdir(new_dir):
             self.source_dir = new_dir
-            self.new_exports = get_recording_dirs(new_dir)
         else:
             logger.warning('"{}" is not a directory'.format(new_dir))
             return
-        if self.new_exports is []:
-            logger.warning('"{}" does not contain recordings'.format(new_dir))
-            return
-
-        self.add_exports()
-        self._update_ui()
 
     def set_dest_dir(self, new_dir):
-        new_dir = new_dir
-
         new_dir = os.path.expanduser(new_dir)
         if os.path.isdir(new_dir):
             self.destination_dir = new_dir
@@ -124,69 +160,41 @@ class Batch_Exporter(Analysis_Plugin_Base):
             logger.warning('"{}" is not a directory'.format(new_dir))
             return
 
-        self.exports = []
-        self.add_exports()
+    def _clear_previous(self):
+        del self.previous_exports[:]
         self._update_ui()
 
-    def add_exports(self):
-        outfiles = set()
-        for d in self.new_exports:
-            logger.debug("Adding new export.")
-            should_terminate = mp.Value(c_bool, False)
-            frames_to_export = mp.Value(c_int, 0)
-            current_frame = mp.Value(c_int, 0)
-            start_frame = None
-            end_frame = None
-            export_dir = d
-            user_dir = self.g_pool.user_dir
+    def _clear_avail(self):
+        del self.available_exports[:]
+        self._update_avail_recs_menu()
 
-            # we need to know the timestamps of our exports.
-            try:
-                frames_to_export.value = len(np.load(os.path.join(export_dir, 'world_timestamps.npy')))
-            except:
-                logger.error('Invalid export directory: {}'.format(export_dir))
-                continue
+    def _update_avail_recs_menu(self):
+        del self.avail_recs_menu[:]
+        if self.available_exports:
+            for avail in self.available_exports:
+                self.avail_recs_menu.append(ui.Switch('selected', avail, label=avail['source']))
+        else:
+            self.avail_recs_menu.append(ui.Info_Text('No recordings available yet. Use Search to find recordings.'))
 
-            # Here we make clones of every plugin that supports it.
-            # So it runs in the current config when we lauch the exporter.
-            plugins = self.g_pool.plugins.get_initializers()
-
-            # make a unique name created from rec_session and dir name
-            rec_session, rec_dir = export_dir.rsplit(os.path.sep, 2)[1:]
-            out_name = rec_session+"_"+rec_dir+".mp4"
-            out_file_path = os.path.join(self.destination_dir, out_name)
-            if out_file_path in outfiles:
-                logger.error("This export setting would try to save {} at least twice please rename dirs to prevent this. Skipping File".format(out_file_path))
-            else:
-                outfiles.add(out_file_path)
-                logger.info("Exporting to: {}".format(out_file_path))
-
-                process = Export_Process(target=export, args=(should_terminate, frames_to_export, current_frame,
-                                                              export_dir, user_dir, self.g_pool.min_data_confidence,
-                                                              start_frame, end_frame, plugins, out_file_path, {}))
-                self.exports.append(process)
-
-    def start(self):
-        self.active_exports = self.exports[:]
-        self.run = True
+    def export_selected(self):
+        for avail in self.available_exports[:]:
+            if avail['selected']:
+                self.available_exports.remove(avail)
+                # self.queued_exports.append()
+        self._update_avail_recs_menu()
 
     def recent_events(self, events):
-        frame = events.get('frame')
-        if not frame:
-            return
-        if self.run:
-            print(self.active_exports)
-            for i in range(len(self.workers)):
-                if self.workers[i] and self.workers[i].is_alive():
-                    pass
-                else:
-                    if self.active_exports:
-                        logger.info("Starting new job")
-                        self.workers[i] = self.active_exports.pop(0)
-                        if not self.workers[i].is_alive():
-                            self.workers[i].start()
-                    else:
-                        self.run = False
+        if self.search_task:
+            recent = [d for d in self.search_task.fetch()]
+            if recent:
+                currently_avail = [rec['source'] for rec in self.available_exports]
+                self.available_exports.extend([{'source': rec, 'selected': True} for rec in recent if rec not in currently_avail])
+                self._update_avail_recs_menu()
+            if self.search_task.completed:
+                self.search_task = None
+                self.search_button.outer_label = ''
+                self.search_button.label = 'Search'
+
 
 def main():
 

--- a/pupil_src/shared_modules/batch_exporter.py
+++ b/pupil_src/shared_modules/batch_exporter.py
@@ -9,6 +9,7 @@ See COPYING and COPYING.LESSER for license details.
 ---------------------------------------------------------------------------~(*)
 '''
 
+import numpy as np
 from pyglui import ui
 import os
 import sys
@@ -85,17 +86,19 @@ class Batch_Exporter(Analysis_Plugin_Base):
         self.menu.append(ui.Text_Input('destination_dir', self, label='Destination Directory', setter=self.set_dest_dir))
         self.menu.append(ui.Button('Export selected', self.export_selected))
         self.menu.append(ui.Button('Clear search results', self._clear_avail))
-        self.menu.append(ui.Separator())
 
         self._update_ui()
 
     def _update_ui(self):
         del self.menu.elements[7:]
 
-        for queued in self.queued_exports[::-1]:
-            def prio_qd():
-                pass
-            self.menu.append(ui.Button('Prioritize', prio_qd, outer_label=queued['dest']))
+        if self.queued_exports:
+            self.menu.append(ui.Separator())
+            self.menu.append(ui.Info_Text('Queued exports:'))
+            for queued in self.queued_exports[::-1]:
+                def prio_qd():
+                    pass
+                self.menu.append(ui.Button('Prioritize', prio_qd, outer_label=queued['dest']))
 
         if self.active_exports:
             self.menu.append(ui.Separator())
@@ -179,9 +182,31 @@ class Batch_Exporter(Analysis_Plugin_Base):
     def export_selected(self):
         for avail in self.available_exports[:]:
             if avail['selected']:
+                try:
+                    frames_to_export = len(np.load(os.path.join(avail['source'], 'world_timestamps.npy')))
+                except:
+                    logger.error('Invalid export directory: {}'.format(avail['source']))
+                    self.available_exports.remove(avail)
+                    continue
+
+                # make a unique name created from rec_session and dir name
+                rec_session, rec_dir = avail['source'].rsplit(os.path.sep, 2)[1:]
+                out_name = rec_session+"_"+rec_dir+".mp4"
+                out_file_path = os.path.join(self.destination_dir, out_name)
+
+                if (out_file_path in (e['dest'] for e in self.queued_exports) or
+                        out_file_path in (e.out_file_path for e in self.active_exports)):
+                    logger.error("This export setting would try to save {} at least twice please rename dirs to prevent this. Skipping recording.".format(out_file_path))
+                    continue
+                if out_file_path in (e.out_file_path for e in self.previous_exports):
+                    logger.error("This export setting would the previous export {}. Please clear the previous exports if you want to overwrite it.".format(out_file_path))
+                    continue
+
+                export = {'source': avail['source'], 'dest': out_file_path, 'frames_to_export': frames_to_export}
                 self.available_exports.remove(avail)
-                # self.queued_exports.append()
+                self.queued_exports.append(export)
         self._update_avail_recs_menu()
+        self._update_ui()
 
     def recent_events(self, events):
         if self.search_task:

--- a/pupil_src/shared_modules/batch_exporter.py
+++ b/pupil_src/shared_modules/batch_exporter.py
@@ -58,20 +58,16 @@ class Batch_Exporter(Analysis_Plugin_Base):
     def __init__(self, g_pool):
         super().__init__(g_pool)
 
-        # initialize empty menu
-        # and load menu configuration of last session
-        self.menu = None
-
         self.exports = []
         self.new_exports = []
         self.active_exports = []
-        default_path = os.path.expanduser('~/work/pupil/recordings/demo')
+        default_path = os.path.expanduser('~/work/pupil/recordings/BATCH')
         self.destination_dir = default_path
         self.source_dir = default_path
 
         self.run = False
-        self.workers = [None for x in range(mp.cpu_count())]
-        logger.info("Using a maximum of {} CPUs to process visualizations in parallel...".format(mp.cpu_count()))
+        self.workers = [None for x in range(mp.cpu_count() - 1)]
+        logger.info("Using a maximum of {} CPUs to process visualizations in parallel...".format(mp.cpu_count() - 1))
 
     def init_ui(self):
         self.add_menu()
@@ -91,9 +87,8 @@ class Batch_Exporter(Analysis_Plugin_Base):
             submenu = ui.Growing_Menu("Export Job {}: '{}'".format(idx, job.out_file_path))
             progress_bar = ui.Slider('Progress', getter=job.status, min=0, max=job.frames_to_export.value)
             progress_bar.read_only = True
-            progress_bar.display_format = '%i frames'
             submenu.append(progress_bar)
-            submenu.append(ui.Button('cancel', job.cancel))
+            submenu.append(ui.Button('Cancel', job.cancel))
             self.menu.append(submenu)
         if not self.exports:
             self.menu.append(ui.Info_Text('Please select a Recording Source directory from which to pull all recordings for the batch export.'))
@@ -180,12 +175,13 @@ class Batch_Exporter(Analysis_Plugin_Base):
         if not frame:
             return
         if self.run:
+            print(self.active_exports)
             for i in range(len(self.workers)):
                 if self.workers[i] and self.workers[i].is_alive():
                     pass
                 else:
-                    logger.info("starting new job")
                     if self.active_exports:
+                        logger.info("Starting new job")
                         self.workers[i] = self.active_exports.pop(0)
                         if not self.workers[i].is_alive():
                             self.workers[i].start()

--- a/pupil_src/shared_modules/batch_exporter.py
+++ b/pupil_src/shared_modules/batch_exporter.py
@@ -51,7 +51,6 @@ def get_recording_dirs(data_dir):
 class Batch_Export(System_Plugin_Base):
     """Sub plugin that manages a single batch export"""
     uniqueness = 'not_unique'
-    menu_icon_order = {'in_queue': 0.8, 'active': 0.7, 'finished': 0.9}
     icon_font = 'pupil_icons'
     icon_chr = chr(0xe2c4)  # character shown in menu icon
 
@@ -77,7 +76,6 @@ class Batch_Export(System_Plugin_Base):
         self.menu[0].label = 'Cancel'
         self.menu_icon.indicator_start = 0.
         self.menu_icon.indicator_stop = 0.1
-        self.menu_icon.order = self.menu_icon_order['in_queue']
         self.menu_icon.tooltip = export_name
         self.menu.append(ui.Text_Input('rec_dir', self, label='Recording', setter=lambda x: None))
         self.menu.append(ui.Text_Input('out_file_path', self, label='Output', setter=lambda x: None))
@@ -88,21 +86,24 @@ class Batch_Export(System_Plugin_Base):
 
     def recent_events(self, events):
         if self.process:
-            recent = [d for d in self.process.fetch()]
-            if recent:
-                self.status, self.progress = recent[-1]
+            try:
+                recent = [d for d in self.process.fetch()]
+            except Exception as e:
+                self.status, self.progress = '{}: {}'.format(type(e).__name__, e), 0
+            else:
+                if recent:
+                    self.status, self.progress = recent[-1]
+
 
             # Update status if process has been canceled or completed
             if self.process.canceled:
                 self.process = None
                 self.status = 'Export has been canceled.'
                 self.notify_all({'subject': 'batch_export.canceled', 'out_file_path': self.out_file_path})
-                self.menu_icon.order = self.menu_icon_order['finished']
                 self.menu[0].label = 'Close'  # change button label back to close
             elif self.process.completed:
                 self.process = None
                 self.notify_all({'subject': 'batch_export.completed', 'out_file_path': self.out_file_path})
-                self.menu_icon.order = self.menu_icon_order['finished']
                 self.menu[0].label = 'Close'  # change button label back to close
 
         if self.in_queue:
@@ -119,20 +120,18 @@ class Batch_Export(System_Plugin_Base):
                 self._accelerate = True
         else:
             self.menu_icon.indicator_start = 0.
-            self.menu_icon.indicator_stop = max(1, self.progress) / self.frames_to_export
+            self.menu_icon.indicator_stop = self.progress / self.frames_to_export
 
     def on_notify(self, n):
         if n['subject'] == 'batch_export.should_start' and n['out_file_path'] == self.out_file_path:
             self.init_export()
         if n['subject'] == 'batch_export.should_cancel' and n.get('out_file_path', self.out_file_path) == self.out_file_path:
             self.cancel_export()
-            self.menu_icon.order = self.menu_icon_order['finished']
             if n.get('remove_menu', False):
                 self.alive = False
 
     def init_export(self):
         self.in_queue = False
-        self.menu_icon.order = self.menu_icon_order['active']
         args = (self.rec_dir, self.g_pool.user_dir, self.g_pool.min_data_confidence,
                 None, None, self.plugins, self.out_file_path, {})
         self.process = bh.Task_Proxy('Pupil Batch Export {}'.format(self.out_file_path), export_function, args=args)
@@ -160,7 +159,7 @@ class Batch_Exporter(Analysis_Plugin_Base):
     icon_chr = chr(0xec05)
     icon_font = 'pupil_icons'
 
-    def __init__(self, g_pool, source_dir='~/work/pupil/recordings/BATCH', destination_dir='~/'):
+    def __init__(self, g_pool, source_dir='~/', destination_dir='~/'):
         super().__init__(g_pool)
 
         self.available_exports = []
@@ -181,7 +180,7 @@ class Batch_Exporter(Analysis_Plugin_Base):
         self.menu.label = 'Batch Export Recordings'
 
         self.menu.append(ui.Info_Text('Search will walk through the source direcotry recursively and detect available Pupil recordings.'))
-        self.menu.append(ui.Text_Input('source_dir', self, label='Source Directory', setter=self.set_src_dir))
+        self.menu.append(ui.Text_Input('source_dir', self, label='Source directory', setter=self.set_src_dir))
 
         self.search_button = ui.Button('Search', self.detect_recordings)
         self.menu.append(self.search_button)
@@ -190,7 +189,7 @@ class Batch_Exporter(Analysis_Plugin_Base):
         self._update_avail_recs_menu()
         self.menu.append(self.avail_recs_menu)
 
-        self.menu.append(ui.Text_Input('destination_dir', self, label='Destination Directory', setter=self.set_dest_dir))
+        self.menu.append(ui.Text_Input('destination_dir', self, label='Destination directory', setter=self.set_dest_dir))
         self.menu.append(ui.Button('Export selected', self.queue_selected))
         self.menu.append(ui.Button('Clear search results', self._clear_avail))
         self.menu.append(ui.Separator())

--- a/pupil_src/shared_modules/exporter.py
+++ b/pupil_src/shared_modules/exporter.py
@@ -53,8 +53,9 @@ class Global_Container(object):
 def export(rec_dir, user_dir, min_data_confidence, start_frame=None, end_frame=None,
            plugin_initializers=(), out_file_path=None, pre_computed={}):
 
-    logger = logging.getLogger(__name__+' with pid: '+str(os.getpid()))
-    start_status = 'Starting video export with pid: {}'.format(os.getpid())
+    PID = str(os.getpid())
+    logger = logging.getLogger(__name__+' with pid: '+PID)
+    start_status = 'Starting video export with pid: {}'.format(PID)
     print(start_status)
     yield start_status, 0
 
@@ -181,7 +182,7 @@ def export(rec_dir, user_dir, min_data_confidence, start_frame=None, end_frame=N
 
             writer.write_video_frame(frame)
             current_frame += 1
-            yield 'Exporting', current_frame
+            yield 'Exporting with pid {}'.format(PID), current_frame
 
         writer.close()
         writer = None

--- a/pupil_src/shared_modules/exporter.py
+++ b/pupil_src/shared_modules/exporter.py
@@ -196,9 +196,10 @@ def export(rec_dir, user_dir, min_data_confidence, start_frame=None, end_frame=N
 
     except GeneratorExit:
         print('Video export with pid {} was canceled.'.format(os.getpid()))
-    except:
+    except Exception as e:
         from time import sleep
         import traceback
         trace = traceback.format_exc()
         print('Process Export (pid: {}) crashed with trace:\n{}'.format(os.getpid(), trace))
+        yield e
         sleep(1.0)

--- a/pupil_src/shared_modules/video_export_launcher.py
+++ b/pupil_src/shared_modules/video_export_launcher.py
@@ -127,11 +127,15 @@ class Video_Export_Launcher(Analysis_Plugin_Base):
 
     def recent_events(self, events):
         for e in self.exports:
-            recent = [d for d in e.fetch()]
-            if recent:
-                e.status, e.progress = recent[-1]
-            if e.canceled:
-                e.status = 'Export has been canceled.'
+            try:
+                recent = [d for d in e.fetch()]
+            except Exception as e:
+                self.status, self.progress = '{}: {}'.format(type(e).__name__, e), 0
+            else:
+                if recent:
+                    e.status, e.progress = recent[-1]
+                if e.canceled:
+                    e.status = 'Export has been canceled.'
 
     def cleanup(self):
         """ called when the plugin gets terminated.

